stdlib/public/core/Integers.swift.gyb

```diff
@@ -283,8 +340,8 @@ public func != <T : BinaryInteger, U : BinaryInteger>(lhs:T, rhs: U) -> Bool {
 
 @_transparent
 public func < <T : BinaryInteger, U : BinaryInteger>(lhs: T, rhs: U) -> Bool {
-  let lhsSign = lhs < 0 ? -1 : lhs > 0 ? 1 : 0
-  let rhsSign = rhs < 0 ? -1 : rhs > 0 ? 1 : 0
+  let lhsSign = lhs < (0 as T) ? -1 : lhs > (0 as T) ? 1 : 0
+  let rhsSign = rhs < (0 as U) ? -1 : rhs > (0 as U) ? 1 : 0
```

Comment:
These are not ambiguous, it's the inlining. It is trying to inline itself. Type cast makes it call the non-generic version, and avoid the problem.
We already have a (Self, Self) comparison requirement on the `BinaryInteger`, it's just called `isLess(than:)`. Given [SE-0091](https://github.com/apple/swift-evolution/blob/master/proposals/0091-improving-operators-in-protocols.md) it will become a `static func < (Self, Self) -> Bool`.
For now, changing explicit type cast to using `isLess(than:)`.